File: Data-Structures-and-Algorithms-using-Swift/Sorting/CountSort.swift
import Foundation

func countSort(_ arr : inout [Int], _ lowerRange : Int, _ upperRange : Int) {
    var j : Int
    let size : Int = arr.count
    let range : Int = upperRange - lowerRange
    var count : [Int] = Array(repeating: 0, count: range)
    var i : Int = 0
    while (i < size) {
        count[arr[i] - lowerRange] += 1
        i += 1
    }
    
    j = 0
    i = 0
    while (i < range) {
        while (count[i] > 0) {
            arr[j] = i + lowerRange
            j += 1
            count[i] -= 1
        }
        i += 1
    }
}

// Testing Code.
var array : [Int] = [23, 24, 22, 21, 26, 25, 27, 28, 21, 21]
countSort( &array,20,30)
print(array)

// [21, 21, 21, 22, 23, 24, 25, 26, 27, 28]

File: Data-Structures-Algorithms-In-TS/Sorting/CountSort.ts
function countSort(arr: number[], lowerRange: number, upperRange: number): void {
    let i = 0, j = 0;
    let size = arr.length;
    let range = upperRange - lowerRange + 1;
    let count = Array(range).fill(0);

    for (i = 0; i < size; i++) {
        count[arr[i] - lowerRange]++;
    }

    for (i = 0; i < range; i++) {
        for (; count[i] > 0; count[i]--) {
            arr[j++] = i + lowerRange;
        }
    }
}

// Testing code.
let array: number[] = [2, 6, 4, 1, 5, 8, 1, 4, 6, 1];
countSort(array, 0, 9);
console.log(JSON.stringify(array));

/*
[1,1,1,2,4,4,5,6,6,8]
*/
